### DISCUSSION TITLE
How to use semaphore to do async rendering?

### DISCUSSION THREAD

**Discussion by SilenSang** [Tencent/libpag#2116] [Q&A] [answered]
The class `BackendSemaphore` is documented as "**Wrapper class for passing into and receiving data from PAG about a backend semaphore object.**", so what the **backend semaphore object** it wraps? Say I have two threads, one is where `PAGPlayer` in, and the other is the rendering thread of another OpenGL context. The two threads share a texture, then how can I use BackendSemaphore to coordinate the drawing of the two threads? How to call `PAGPlayer::wait` and `PAGPlayer::flushAndSignalSemaphore`? Is C++ `std::mutex` a kind of semaphore? Can you show some sample code to illustrate this concept? Thanks in advance.

**Comment by domchen** [comment, marked as answer]
The BackendSemaphore currently wraps the GLSync object. Unlike C++ std::mutex, GLSync serves a distinct purpose. For assistance in using the GLSync object, please google the sample code for `glWaitSync()` and `glFenceSync()` in the OpenGL APIs. The corresponding PAG APIs works in the same manner as those in OpenGL.
